### PR TITLE
gh-actions: test LoongArch64

### PR DESCRIPTION
**Pull request by XiWeiGu**

### CODE REVIEW

File: docker/cross-files/loongarch64-gcc-7.cross [hunk 1]
```diff
@@ -0,0 +1,18 @@
+[binaries]
+c = 'loongarch64-unknown-linux-gnu-gcc'
+cpp = 'loongarch64-unknown-linux-gnu-g++'
+ar = 'loongarch64-unknown-linux-gnu-ar'
+strip = 'loongarch64-unknown-linux-gnu-strip'
+objcopy = 'loongarch64-unknown-linux-gnu-objcopy'
+ld = 'loongarch64-unknown-linux-gnu-ld'
+exe_wrapper = ['qemu-loongarch64', '-L', '/opt/cross-tools/target/usr']
```
Comment (mr-c) [inline docker/cross-files/loongarch64-gcc-7.cross:8]:
```suggestion
exe_wrapper = ['qemu-loongarch64', '-cpu', 'la464-loongarch-cpu', '-L', '/opt/cross-tools/target/usr']
```
Might fix the test running errors

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -545,3 +545,36 @@ jobs:
...
+        export PATH=/opt/cross-tools/bin:$PATH
+        export LD_LIBRARY_PATH=/opt/cross-tools/target/usr/lib64:$LD_LIBRARY_PATH
+        export LD_LIBRARY_PATH=/opt/cross-tools/loongarch64-unknown-linux-gnu/lib64:$LD_LIBRARY_PATH
+        ninja -C build -v
+    - name: Test
+      run: |
+        export PATH=$GITHUB_WORKSPACE:/opt/cross-tools/bin:$PATH
+        export LD_LIBRARY_PATH=/opt/cross-tools/target/usr/lib64:$LD_LIBRARY_PATH
+        export LD_LIBRARY_PATH=/opt/cross-tools/loongarch64-unknown-linux-gnu/lib64:$LD_LIBRARY_PATH
+        ninja -C build -v test
```
Comment (mr-c) [inline .github/workflows/ci.yml:580]:
```suggestion
        meson test -C build --timeout-multiplier 0 --print-errorlogs
```

**Review by mr-c — CHANGES_REQUESTED** [review]
Thanks!

In `.github/workflows/ci.yml`, can you replace the `export ...` environment variable setting lines with a shared `env` section? Example syntax (but you don't need the CFLAGS as you are using the cross-file): https://github.com/simd-everywhere/simde/blob/554ab18acf70fef7f93b043a46d70942c035be80/.github/workflows/ci.yml#L93

The repo https://github.com/loongson/build-tools is `archived` (closed). Is there an alternative that will be maintained? Also, that repo is missing the source code, which is a GPL violation

File: .github/workflows/ci.yml (hunk 2 above)
Comment (XiWeiGu) [reply .github/workflows/ci.yml:580]: I will take your suggestion into consideration and implement it. Thank you for taking the time to review.

**Comment by XiWeiGu** [comment]
> Thanks!
> 
> In `.github/workflows/ci.yml`, can you replace the `export ...` environment variable setting lines with a shared `env` section? Example syntax (but you don't need the CFLAGS as you are using the cross-file):
> 
> https://github.com/simd-everywhere/simde/blob/554ab18acf70fef7f93b043a46d70942c035be80/.github/workflows/ci.yml#L93
> 
> The repo https://github.com/loongson/build-tools is `archived` (closed). Is there an alternative that will be maintained? Also, that repo is missing the source code, which is a GPL violation

Thank you for your feedback. I have consulted with the employees of Loongson, and they have informed me that they are currently in talks about uploading the source code. Hopefully, they will resolve the issue soon.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -561,3 +561,30 @@ jobs:
...
+    - name: set env
+      run: |
+        echo "LD_LIBRARY_PATH=/opt/cross-tools/target/usr/lib64:/opt/cross-tools/loongarch64-unknown-linux-gnu/lib64:$LD_LIBRARY_PATH" >> $GITHUB_ENV
+        echo "PATH=$GITHUB_WORKSPACE:/opt/cross-tools/bin:$PATH" >> $GITHUB_ENV
+    - name: Configure
+      run: meson setup build --cross-file=docker/cross-files/loongarch64-gcc-7.cross
+    - name: Build
+      run: ninja -C build -v
+    # Due to the qemu-loongarch64(v2022.09.06) provided by Loongson causing timeouts in four sets of test cases,
+    # the tests have been temporarily disabled.
```
Comment (mr-c) [inline .github/workflows/ci.yml:588]: Okay, we can skip just those tests that timeout; let me know which tests are timing out and I'll make the changes for you

Comment (XiWeiGu) [reply .github/workflows/ci.yml:588]: > Summary of Failures:
> 
> 1547/1550 x86/svml/emul/c                         TIMEOUT        30.02s   killed by signal 15 SIGTERM
> 1548/1550 x86/svml/emul/cpp                       TIMEOUT        30.01s   killed by signal 15 SIGTERM
> 1549/1550 x86/svml/native/c                       TIMEOUT        30.02s   killed by signal 15 SIGTERM
> 1550/1550 x86/svml/native/cpp                     TIMEOUT        30.03s   killed by signal 15 SIGTERM
> 
> Ok:                 1546
> Expected Fail:      0
> Fail:               0
> Unexpected Pass:    0
> Skipped:            0
> Timeout:            4

Comment (XiWeiGu) [reply .github/workflows/ci.yml:588]: All test cases have been successfully passed when running them on the native 3A5000 machine.

**Review by mr-c — CHANGES_REQUESTED** [review]
This looks good to me. Thank you @XiWeiGu ;

What's the status of GPL compliance ("complete corresponding source code") for the Loongson provided binaries of Qemu & GCC?

**Comment by GuXiWei** [comment]
Yes, as of now, GCC has already uploaded the corresponding source code.
The status of Qemu's GPL compliance is still under negotiation and the process is ongoing. Thank you for your understanding.

**Comment by XiWeiGu** [comment]
Yes, as of now, GCC has already uploaded the corresponding source code.
The status of Qemu's GPL compliance is still under negotiation and the process is ongoing. Thank you for your understanding.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -644,3 +644,34 @@ jobs:
...
+    - name: CPU Information
+      run: cat /proc/cpuinfo
+    - name: Install APT Dependencies
+      run: sudo apt-get install -y ninja-build ninja-build meson
+    - name: loongarch64-toolchain
+      run: |
+        wget https://github.com/loongson/build-tools/releases/download/2022.09.06/loongarch64-clfs-7.3-cross-tools-gcc-glibc.tar.xz
+        tar -xvf loongarch64-clfs-7.3-cross-tools-gcc-glibc.tar.xz -C /opt
+        wget https://github.com/loongson/build-tools/releases/download/2022.09.06/qemu-loongarch64
+        chmod +x qemu-loongarch64
```
Comment (mr-c) [inline .github/workflows/ci.yml:661]: https://github.com/loongson/build-tools#readme states that the sourcecode for this is https://download.qemu.org/qemu-8.0.0.tar.xz ; would it work to install http://deb.debian.org/debian/pool/main/q/qemu/qemu-user-static_8.0+dfsg-4_amd64.deb ?
```suggestion
        wget http://deb.debian.org/debian/pool/main/q/qemu/qemu-user-static_8.0+dfsg-4_amd64.deb
        apt install ./qemu-user-static_8.0+dfsg-4_amd64.deb
```

**Review by mr-c — APPROVED** [review]
Thank you @XiWeiGu 

I got Qemu 7.2 to work, also skipping the svml tests.

I'm going to queue https://github.com/simd-everywhere/simde/commit/eff59f23305532ffc9971955ffcea2d6b7cc8c2c for merging

**Comment by mr-c** [comment]
Merged as https://github.com/simd-everywhere/simde/commit/62b9f2a222fb27e5a6ead7dff541f493f0cda8f7

**Pull request closed without merging**